File: examples/service.scala
package examples

import scala.collection.mutable

/** Thread-safe in-memory cache with configurable eviction. */
class Cache[K, V](maxSize: Int) {

  private val store = mutable.LinkedHashMap.empty[K, V]
  private val lock = new Object

  /** Get a value by key. Returns None if missing or evicted. */
  def get(key: K): Option[V] = lock.synchronized {
    store.get(key).filter { _ =>
      // Move to end (most recently used)
      store.remove(key)
      store.put(key, _)
      true
    }
  }

  /** Insert a key-value pair. Evicts oldest entry if at capacity. */
  def put(key: K, value: V): Unit = lock.synchronized {
    if (store.size >= maxSize && !store.contains(key)) {
      store.remove(store.head._1)
    }
    store.put(key, value)
  }

  /** Remove a specific key. */
  def remove(key: K): Unit = lock.synchronized {
    store.remove(key)
  }

  /** Clear all entries. */
  def clear(): Unit = lock.synchronized {
    store.clear()
  }

  /** Current number of entries. */
  def size: Int = lock.synchronized store.size
}

object Cache {
  def apply[K, V](maxSize: Int): Cache[K, V] = new Cache[K, V](maxSize)
}

object Main extends App {
  val cache = Cache[String, Int](100)
  cache.put("one", 1)
  cache.put("two", 2)
  cache.put("three", 3)

  cache.get("one").foreach(v =>
    println(s"Found: $v")
  )

  // TODO: Add async support
  // FIXME: Add serialization
}

File: examples/service.java
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Thread-safe in-memory cache with TTL support.
 */
public class Cache<K, V> {

    private final Map<K, Entry<V>> store;
    private final int maxSize;
    private final long defaultTtlMs;

    private static class Entry<V> {
        final V value;
        final long expiresAt;

        Entry(V value, long ttlMs) {
            this.value = value;
            this.expiresAt = System.currentTimeMillis() + ttlMs;
        }

        boolean isExpired() {
            return System.currentTimeMillis() > expiresAt;
        }
    }

    public Cache(int maxSize, long defaultTtlMs) {
        this.maxSize = maxSize;
        this.defaultTtlMs = defaultTtlMs;
        this.store = new ConcurrentHashMap<>();
    }

    public Optional<V> get(K key) {
        Entry<V> entry = store.get(key);
        if (entry == null || entry.isExpired()) {
            store.remove(key);
            return Optional.empty();
        }
        return Optional.of(entry.value);
    }

    public void put(K key, V value) {
        if (store.size() >= maxSize) {
            evictExpired();
            if (store.size() >= maxSize) {
                store.remove(store.keySet().iterator().next());
            }
        }
        store.put(key, new Entry<>(value, defaultTtlMs));
    }

    public void remove(K key) {
        store.remove(key);
    }

    public void clear() {
        store.clear();
    }

    private void evictExpired() {
        store.entrySet().removeIf(e -> e.getValue().isExpired());
    }

    public static void main(String[] args) {
        Cache<String, Integer> cache = new Cache<>(100, 60_000);
        cache.put("key1", 42);
        cache.put("key2", 100);

        cache.get("key1").ifPresent(v ->
            System.out.printf("Found: %d%n", v)
        );

        // TODO: Add stats tracking
        // FIXME: WeakReference keys for large caches
    }
}
